From 4d8620c60c9cda15973901d10d1f5fd8bc94d312 Mon Sep 17 00:00:00 2001
From: startergo <7897244+startergo@users.noreply.github.com>
Date: Wed, 24 Jun 2026 10:03:13 -0500
Subject: [PATCH 15/36] [sl-port-610] Fix collapsed glyph advances on 10.6
 (squished-text regression): platformWidthForGlyph used
 CTFontGetUnsummedAdvancesForGlyphsAndStyle (10.12+, returns 0 on 10.6 -> all
 glyphs ~0 width). Gate on __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ <
 101200, use CTFontGetAdvancesForGlyphs (10.6-available)

---
 .../WebCore/platform/graphics/cocoa/FontCocoa.mm   | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/Source/WebCore/platform/graphics/cocoa/FontCocoa.mm b/Source/WebCore/platform/graphics/cocoa/FontCocoa.mm
index c94b873a30..458298641f 100644
--- a/Source/WebCore/platform/graphics/cocoa/FontCocoa.mm
+++ b/Source/WebCore/platform/graphics/cocoa/FontCocoa.mm
@@ -628,6 +628,7 @@ float Font::platformWidthForGlyph(Glyph glyph) const
     bool horizontal = platformData().orientation() == FontOrientation::Horizontal;
     CGFontRenderingStyle style = kCGFontRenderingStyleAntialiasing | kCGFontRenderingStyleSubpixelPositioning | kCGFontRenderingStyleSubpixelQuantization | kCGFontAntialiasingStyleUnfiltered;
 
+#if !(defined(__ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__) && __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ < 101200)
     if (platformData().size()) {
         CTFontOrientation orientation = horizontal || m_isBrokenIdeographFallback ? kCTFontOrientationHorizontal : kCTFontOrientationVertical;
         // FIXME: Remove this special-casing when <rdar://problem/28197291> and <rdar://problem/28662086> are fixed.
@@ -636,6 +637,19 @@ float Font::platformWidthForGlyph(Glyph glyph) const
         else
             CTFontGetUnsummedAdvancesForGlyphsAndStyle(m_platformData.ctFont(), orientation, style, &glyph, &advance, 1);
     }
+#else
+    // [leopard] CTFontGetUnsummedAdvancesForGlyphsAndStyle is 10.12+; on 10.6 it
+    // is unbound and returns a zero advance, collapsing every glyph to ~0 width
+    // (illegible overlapping text on all pages). CTFontGetAdvancesForGlyphs is
+    // available on 10.6 and returns correct advances; use it directly. (610
+    // dropped 605's canUseFastGlyphAdvanceGetter/CGFont fast path, so don't
+    // depend on it here.)
+    UNUSED_PARAM(style);
+    if (platformData().size()) {
+        CTFontOrientation orientation = horizontal || m_isBrokenIdeographFallback ? kCTFontOrientationHorizontal : kCTFontOrientationVertical;
+        CTFontGetAdvancesForGlyphs(m_platformData.ctFont(), orientation, &glyph, &advance, 1);
+    }
+#endif
     return advance.width + m_syntheticBoldOffset;
 }
 
-- 
2.54.0

